proof_ipa.go

```diff
@@ -347,22 +386,37 @@ func DeserializeProof(vp *VerkleProof, statediff StateDiff) (*Proof, error) {
 
 	// turn statediff into keys and values
 	for _, stemdiff := range statediff {
-		for _, suffixdiff := range stemdiff.SuffixDiffs {
+		for _, upt := range stemdiff.Updates {
 			var k [32]byte
 			copy(k[:StemSize], stemdiff.Stem[:])
-			k[StemSize] = suffixdiff.Suffix
+			k[StemSize] = upt.Suffix
 			keys = append(keys, k[:])
-			if suffixdiff.CurrentValue != nil {
-				prevalues = append(prevalues, suffixdiff.CurrentValue[:])
-			} else {
-				prevalues = append(prevalues, nil)
-			}
-
-			if suffixdiff.NewValue != nil {
-				postvalues = append(postvalues, suffixdiff.NewValue[:])
-			} else {
-				postvalues = append(postvalues, nil)
-			}
+			prevalues = append(prevalues, upt.Current[:])
+			postvalues = append(postvalues, upt.New[:])
+		}
+		for _, ins := range stemdiff.Inserts {
+			var k [32]byte
+			copy(k[:StemSize], stemdiff.Stem[:])
+			k[StemSize] = ins.Suffix
+			keys = append(keys, k[:])
+			prevalues = append(prevalues, nil)
+			postvalues = append(postvalues, slices.Clone(ins.New[:]))
```

Comment:
This, and line 410, are the fixes for the test that was broken.

Reply:
Weird, so this means that `stemdiff` is somewhat mutated after affecting the returned `Proof`?

Reply:
the area that is used to store the location of the iterator is reused over and over, and if we pass a reference to it, it will change as the iterator progress. This is not the first time we fall for this, and probably not the last time either, it's quite subtle and easy to forget.